Mathlib/Analysis/Convex/Halfspace.lean

```diff
@@ -0,0 +1,126 @@
+/-
+Copyright (c) 2024 Hyeokjun Kwon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Hyeokjun Kwon
+-/
+import Mathlib.Analysis.InnerProductSpace.Dual
+import Mathlib.Analysis.NormedSpace.AddTorsorBases
+
+/-!
+# Halfspace
+
+This file defines halfspaces in an real vector space. Halfspaces are sets of points that lie on one
+side of a hyperplane including the hyperplane itself. We define halfspaces using a normalized
+vector, direction, and a scalar, bound. It is formalised as a set of points that the inner product
+with the direction vector is less than or equal to the bound.
+
+Equivalently, halfspaces can be defined as the preimage of a closed interval under the dual linear
+map of the direction vector.
+-/
+
+open LinearMap Set
+open scoped BigOperators Convex Pointwise
+
+
+/-- Halfspace is a set of points that lie on one side of a hyperplane.
+  or a set of points that their inner product with the direction vector is less than the bound -/
+structure Halfspace (E : Type) [NormedAddCommGroup E] [InnerProductSpace ℝ E] where
```

Comment:
Do we need `InnerProductSpace` for this? Note that, e.g., `Real × Real` is not an inner product space. In any space, we can take a dual vector. Also, please use `Type*`, not `Type`.